src/wasm-traversal.h

```diff
@@ -157,185 +149,209 @@ struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
     self->visitMemory(&module->memory);
     self->visitModule(module);
   }
-};
 
-template<typename ParentType>
-struct ChildWalker : public WasmWalkerBase<ChildWalker<ParentType>> {
-  ParentType& parent;
+  // Walk implementation. We don't use recursion as ASTs may be highly
+  // nested.
 
-  ChildWalker(ParentType& parent) : parent(parent) {}
+  // Tasks receive the this pointer and a pointer to the pointer to operate on
+  typedef void (*TaskFunc)(SubType*, Expression**);
```

Comment:
How about making this a `std::function`? Also, it might be nice to have it return an Expression* instead of using the class-scoped variable. I think having that explicit would be a little clearer, although I guess it would complicate the default implementations of `doVisitFoo()`... What happens in that case if `visitExpression` and `visitFoo` both want to replace the node? the later one wins?